pkg/distributor/ha_tracker.go

```diff
@@ -0,0 +1,296 @@
+// Wraps updating the timestamp for an replica in the local cache so that we can grab the
+// lock and then check if we should actually update it. The timestamp should only ever go up.
+func (c *tracker) updateTimestamp(cluster, replica string, ts time.Time) {
+	c.electedLock.Lock()
+	defer c.electedLock.Unlock()
+	v := c.elected[cluster]
+	if v.timestamp.Before(ts) {
+		c.elected[cluster] = replicaTracker{
+			replica:   replica,
+			timestamp: ts,
+		}
+	}
+}
+
+func (c *tracker) setReplica(cluster, replica string, ts time.Time) {
+	c.electedLock.Lock()
+	defer c.electedLock.Unlock()
+	// Check if we're changing the replica or just the timestamp.
+	if c.elected[cluster].replica != replica {
+		level.Info(util.Logger).Log("msg", "changed which HA replica we're accepting samples from", "cluster", cluster, "replica", replica, "timestamp", ts, "previous", c.elected[cluster].replica)
+	}
+	c.elected[cluster] = replicaTracker{
+		replica:   replica,
+		timestamp: ts,
+	}
+}
+
+// LookupReplica returns a pointer to the replicaTracker for the replica we should be accepting
+// samples from based on the given cluster. It will return nil if there's any error talking to the KVStore.
+// TODO: this should probably be refactored so that it doesn't modify the values stored in c.elected.
+func (c *tracker) LookupReplica(ctx context.Context, cluster, replica string) *replicaTracker {
+	now := mtime.Now()
+	c.electedLock.RLock()
+	v, ok := c.elected[cluster]
+	c.electedLock.RUnlock()
+
+	// No replica in local cache for this HA cluster yet.
+	if !ok {
+		return c.checkKVStore(ctx, cluster, replica, now)
+	}
+
+	// The replica was in the distributors cache.
+	accept, write := c.checkLocal(ctx, v, cluster, replica, now)
+	if !accept {
+		return &v
+	}
+	if write {
+		err := c.casWrapper(ctx, cluster, replica, now)
+		// If it's an actual error.
+		if _, ok := err.(*ErrDontOverwrite); !ok && err != nil {
+			level.Error(c.logger).Log("msg", "non ErrDontOverwrite error", "err", err)
+			return nil
+		}
+		// If we didn't error then the CAS succeeded and we can set the new replica in the cache.
+		c.setReplica(cluster, replica, now)
+		c.electedLock.RLock()
+		v = c.elected[cluster]
+		c.electedLock.RUnlock()
+		return &v
+	}
+	return &v
+}
+
+// Only calls return false to bail out early, otherwise continues on to return true at the end.
+func (c *tracker) checkKVStore(ctx context.Context, cluster, replica string, now time.Time) *replicaTracker {
+	ret := replicaTracker{}
+	err := c.casWrapper(ctx, cluster, replica, now)
+	if err != nil {
+		// If it's an actual error.
+		if _, ok := err.(*ErrDontOverwrite); !ok {
+			level.Error(c.logger).Log("msg", "non ErrDontOverwrite error", "err", err)
+			return nil
+		}
+	}
+	// There was no error or error was ErrDontOvewrite, so we still want to accept the sample.
+	c.setReplica(cluster, replica, now)
+	c.electedLock.RLock()
+	ret = c.elected[cluster]
+	c.electedLock.RUnlock()
```

Comment:
As discussed: lets not update the local cache in this case - WatchPrefix might be behind for instance, and will clobber this.  Just rely on watch prefix.